lib/ts/MemArena.h

```diff
@@ -0,0 +1,134 @@
+    newer generation, copy the contents of the previous generation to the new generation, and then thaw() the previous generation.
+    Note that coalescence must be done by the caller because MemSpan will only give a reference to the underlying memory.
+ */
+class MemArena
+{
+public:
+  /** Simple internal arena block of memory. Maintains the underlying memory.
+   */
+  struct MemArenaBlock {
+    void *data;
+    size_t size;
+    size_t allocated;
+    MemArenaBlock *next;
+
+    void alloc(size_t n);
+    void free(); ///< Need to free() underlying memory before delete.
+  };
+
+  MemArena();
+  MemArena(size_t n);
+
+  ~MemArena();
+
+  /** MemSpan alloc(size_t n)
+
+      Returns a span of memory within the arena. alloc() is self expanding but DOES NOT self coalesce. This means
+      that no matter the arena size, the caller will always be able to alloc() @a n bytes.
+
+      @param n number of bytes to allocate.
+      @return a MemSpan of the allocated memory.
+      @note in order to guarantee n bytes of memory, the implementation keeps creating new
+              generations until there is a large enough memory chunk. If you need a large chunk of memory,
+              you should be creating a large MemArena and then allocating from it. Rather than relying on the
+              default size and arena resizing itself.
+   */
+  MemSpan alloc(size_t n);
+
+  /** MemArena& freeze(size_t n = 0)
+
+      Will "freeze" a generation of memory. Any memory previously allocated can still be used. This is an
+      important distinction as freeze does not mean that the memory is immutable, only that subsequent allocations
+      will be in a new generation.
+
+      @param n Number of bytes for new generation.
+        if @a n == 0. double in size.
+      @return @c *this
+   */
+  MemArena &freeze(size_t n = 0);
+
+  /** MemArena& thaw()
+
+      Will "thaw" away any previously frozen generations. Any generation that is not the current generation is considered
+      frozen because there is no way to allocate in any of those memory blocks. thaw() is the only mechanism for deallocating
+      memory in the arena (other than destroying the arena itself). Thawing away previous generations means that all spans
+      of memory allocated in those generations are no longer safe to use.
+
+      @return @c *this
+   */
+  MemArena &thaw();
+
+  /// @returns the current generation @c size.
+  size_t
+  size() const
+  {
+    return arena_size;
+  }
+
+  /// @returns the @c remaining space within the generation.
+  size_t
+  remaining() const
+  {
+    return generation->size - generation->allocated;
+  }
+
+  /// @return a @c true if @ptr is in memory owned by this arena, @c false if not.
+  bool contains(void *ptr) const;
+
+private:
+  void create_new_generation();
+
+  const unsigned int DEFAULT_BLOCK_SIZE = 32000; ///< 32kb
```

Comment:
`static constexpr size_t DEFAULT_BLOCK_SIZE = 1<<15;`. That might be a bit large to start.